### ISSUE TITLE
Access to values with FieldArray

### ISSUE THREAD

**Issue by neverfox** [erikras/redux-form#1448]
WIth the v6 API, there is no access to the values of the field array members (just names) so that makes it impossible to do value-conditional logic for the array methods (e.g. remove if the value of the last element is blank etc.) which we were easily doing with v5 (because we had all the values as props). Using a selector seemed like a possibility but that doesn't seem to play well with trying to be DRY because you have to know what art of state you're working with to access the right selector prop rather than just taking the fields array. Does it make sense to let the API serve values for the members?

**Comment by eaardal** [comment]
This is very important functionality we're missing (or can't figure out how to do) in our project here aswell. We need to read several field values and their meta info and then somewhere else in the dom, show an error not connected/related to a input element or Field component. It's more of a virtual field that only displays based on the state of other input fields.

**Comment by cwallervand** [comment]
Absolutely agree, the v6 API need to support conditional validation of fields based on the value of one or several other fields. We recently tried migrating from v5 to v6 and migration stopped when we had to validate the value of one field based on another because we could not find a way to do this.

**Comment by clayne11** [comment]
You can do it using [formValueSelector](http://redux-form.com/6.0.0-rc.4/docs/api/FormValueSelector.md/). Simply grab the value you need to validate against and pass it into the component that needs it.

**Comment by cwallervand** [comment]
Yeah, I saw that in the API, but it also says it will cause the form to
rerender for every change in the form so it shouldnot be used unless
necessary

Den tirsdag 9. august 2016 skrev Curtis Layne <notifications@github.com>
følgende:

**Comment by clayne11** [comment]
That's not necessarily true. It will only cause the form to re-render if you `connect` up the value to the top-level form component. You could also use `formValueSelector` to pipe the value directly into the component that needs the value, then only it will re-render.

I'll admit there is a fair bit of boilerplate required to set this up but it's the best solution we have at the moment. If you use this pattern a lot you could factor out the logic into an HOC.

**Comment by naw** [comment]
To facilitate this approach, could I propose that the form name be added as a prop that `FieldArray` exposes to the underlying component?  This way you can use `formValueSelector` without hardcoding the form name and/or without having to pass the form name all the way down from the top?

**Comment by naw** [comment]
Going beyond that, it seems that both `reduxForm` and `FieldArray` could expose a ready-to-go `formValueSelector` as a prop, so that the wrapped component(s) can consume it directly in an underlying `connect` call rather than manually creating it using the form name. Thoughts?

**Comment by cwallervand** [comment]
I am curious if @erikras has any thoughts about this issue. There should be an easier way to work with multiple fields's data in one place, rather than work around the issue.

**Comment by clayne11** [comment]
@naw that seems like a reasonable idea to me.

**Comment by naw** [comment]
In v5 the whole form was subscribed to all of the values, making it easy to work with values at the expense of poor performance for large forms.

In v6 the whole form is subscribed to no values, making it performant at the expense of convenience when you need to get your hands on values.

You can think of `FieldArray` as having the exact same trade-offs, just on a smaller scale. If `FieldArray` automatically subscribed to all values for you by default, you could still have performance issues (like in v5) if you have a large sub-form for each of your array elements. So this isn't a good option (by default).

The trade-off is performance vs. convenience.

I think there are two pain points here:

1. The sub-forms rendered for a `FieldArray` are probably smallish more often than not. This means for many people, it's probably more helpful to optimize for convenience rather than performance. 

2. `formValueSelector` API is somewhat cumbersome to use (see my suggestions above)

**One way to keep performance for those you need it, but also provide convenience for those who want it**, would be adding a configuration prop to `FieldArray` that lets you opt-in to some or all values (effectively wiring up `formValueSelector` _for you_ so you don't have to do it manually).

The same thing could be done for `reduxForm` at the top-level (i.e. the ability to pass it a configuration prop that subscribes to some or all values, so you don't have to manually wire up `formValueSelector`).   Of course we could write such a thing in user land, but having it built-in to `FieldArray` and `reduxForm` seems more polished.

Thoughts?

**Comment by clayne11** [comment]
There was a discussion about this issue a couple months ago in https://github.com/erikras/redux-form/issues/841.

@erikras's solution was allowing a `Field` to subscribe to multiple values:

```javascript
<Field name={['a', 'b']} ... />
```

This doesn't necessarily solve the issue of subscribing to fields on a form level, though, and it also doesn't help with `FieldArray`.

A solution that doesn't require modifying the `Field`, `FieldArray`, and `reduxForm` components is to include a wrapper component that can pass down an array of field values to it's child components as props.

Something like this:

```javascript
import { Children, cloneElement, PropTypes } from 'react'
import { connect } from 'react-redux'

const WithFields = ({
  names,
  children
}, {
  _reduxForm: {
    selector
  }
}) => (
  connect(
    (state) => selector(state, names)
  )(Children.only(children))
)

WithFields.propTypes = {
  names: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  children: PropTypes.node.isRequired
}

WithFields.contextType = {
  _reduxForm: PropTypes.object.isRequired
}
```

This would also requiring passing down a "pre-bound" `formValueSelector` instance from the form.

**Comment by naw** [comment]
@clayne11 yes, good ideas.

I agree something like `<WithFields>` would work OK.  (Note: you probably know this, but since you can get the form name from `_reduxForm`, you don't actually need the selector to be passed down for this to work, although it would be nice).

Also, if `<Field>` could subscribe to multiple field names (as you mentioned was suggested in #841), that's really the same thing as `<WithFields>`, except in addition to values, you _also_ get the other things like `onChange`.  Both of these approaches _does_ handle the  `<FieldArray>` and `form` cases.  Anywhere you would put `<WithFields>`, you could put a `<Field>` instead.  There's no reason you can't have multiple  `<Field>` components for the same field name --- one used to get values for validation/conditionals within the form, the other used to drive the actual DOM input behavior.

All of these really boil down to the same thing --- it's just a matter of where you specify the names of the fields you want to subscribe to.  Personally I see it as a _both_/_and_ kind of thing -- a `<WithFields>` wrapper is great, but it'd also be nice to be able to subscribe directly in a `FieldArray` or `reduxForm`.  You could implement that functionality internally by injecting a `<WithFields>` wrapper if a certain config prop is passed.

Clearly, all of this can be implemented in user land today without much fuss, so really it's just a question of what kinds of conveniences would be helpful to have in the official API (which of course is subjective).

This is just brainstorming, so there may be more elegant solutions we haven't stumbled on yet. Ultimately, any or all of these ideas seem like a step in the right direction.

**Comment by clayne11** [comment]
I think some solution for dependent fields within `redux-form` itself is definitely a necessary convenience as this is a very common use case.

Also you need more than just the form name to create a `formValueSelector` - you need the `getState` method as well, since not everyone stores the `redux-form` reducer at the `form` key and you also need to account for `ImmutableJS`.

I agree we're stepping in the right direction though. I'd like to see some input from @erikras and then someone can move forward with implementing whatever behavior we decide upon.

**Comment by naw** [comment]
As a point of clarification, `connect` is what is injecting state in _both_ cases, so form name is adequate.

**Comment by naw** [comment]
Somehow `<Field>` is getting what it needs from context, so `WithFields` should be able to get what it needs, too. Right?

**Comment by clayne11** [comment]
That's a good point. So yeah, it's available. It wouldn't be difficult to add this functionality.

**Comment by cwallervand** [comment]
Good ideas from @naw and @clayne11, and I think they will take us in the correcet dircetion. With the proposed ideas, would it be possible to register a "hidden" field in the form, i.e. say a value called `amount` which is a representative field for two other fields: `dollars` and `cents`. 
It should be possible to work with the `amount` field as any other fields, but the field does not need to be rendered to screen.

**Comment by webmobiles** [comment]
for  me it's simple , i've just to get  the value from the  state, looking for name form and input name, fieldArray or another field is the same.
then creating onChange events in inputs, pointing in a global function for the form that can be react  to different valus, to validate or set another values.


```

export const getInputValue = (xinputName = '' ) => {
  if (xinputName  ==='') {
    inputName = props.input.name;
  } else {
    inputName = xinputName;
  }
  if (props.myState && typeof props.myState.form !== 'undefined'
    && typeof props.myState.form[nameForm] !== 'undefined'
    && typeof props.myState.form[nameForm].values !== 'undefined'
  ) {
    let valueReal = props.myState.form[nameForm].values[inputName];
...
```
